src/transformers/commands/serving.py

```diff
@@ -98,131 +135,193 @@ def register_subcommand(parser: ArgumentParser):
+        payload = {
+            "object": "chat.completion.chunk",
+            "id": request_id,
+            "created": int(time.time()),
+            "model": self.args.model_name_or_path,
+            "system_fingerprint": "",
+            "choices": [
+                {
+                    "index": 0,
+                    "delta": {"role": "assistant", "content": content, "tool_calls": []},
+                    "logprobs": None,
+                    "finish_reason": finish_reason,
+                }
+            ],
+        }
+        return f"data: {json.dumps(payload)}\n\n"
+
+    def run(self):
+        app = FastAPI()
+
+        if self.use_continuous_batching:
+            self.continuous_batching(app)
+        else:
+            self.generate(app)
+
+        uvicorn.run(app, host=self.args.host, port=self.args.port, log_level=self.args.log_level)
+
+    def continuous_batching(self, app):
+        generation_config = GenerationConfig(
+            eos_token_id=self.tokenizer.eos_token_id,
+            pad_token_id=self.tokenizer.pad_token_id,
+            use_cache=False,
+            num_blocks=1,
+            block_size=1024,
+            do_sample=False,
+            max_batch_tokens=10,
+            scheduler="fifo",
         )
-        serve_parser.add_argument("--host", type=str, default="localhost", help="Interface the server will listen on.")
-        serve_parser.add_argument("--port", type=int, default=8888, help="Port the serving will listen to.")
-        serve_parser.add_argument("--workers", type=int, default=1, help="Number of http workers")
-        serve_parser.add_argument("--model", type=str, help="Model's name or path to stored model.")
-        serve_parser.add_argument("--config", type=str, help="Model's config name or path to stored model.")
-        serve_parser.add_argument("--tokenizer", type=str, help="Tokenizer name to use.")
-        serve_parser.add_argument(
-            "--device",
-            type=int,
-            default=-1,
-            help="Indicate the device to run onto, -1 indicates CPU, >= 0 indicates GPU (default: -1)",
+
+        manager: ContinuousBatchingManager = self.model.init_continuous_batching(
+            generation_config=generation_config, streaming=True
         )
-        serve_parser.set_defaults(func=serve_command_factory)
+        manager.start()
+
+        @app.post("/v1/chat/completions")
+        def _serve(req: ChatCompletionInput):
+            if not req.stream:
+                return {"error": "Only streaming mode is supported."}
+
+            chat = req.messages
+
+            inputs = self.tokenizer.apply_chat_template(
+                chat, return_tensors="pt", add_generation_prompt=True, tools=req.tools
+            ).to(self.model.device)
+
+            generation_config = GenerationConfig(**(req.generation_config or {}))
+
+            def stream_response(_inputs):
```

Comment:
You might want to add a try/except in this method to return a proper error if anything happens. In "sync" HTTP calls, an error would lead to a HTTP 500. Not ideal but at least the user knows something bad happened. When using streamed requests, the user will get a "connection abruptly closed" but still a HTTP 200 (since headers are returned _before_ the error occurs). This is not always well handled by clients.

For instance TGI returns a chunk with error status + message in such a case `data: {"error": ...}`. Handled [here](https://github.com/huggingface/huggingface_hub/blob/28554021b7655a7d844c32e423d18a5ee1845599/src/huggingface_hub/inference/_common.py#L289) by the InferenceClient.

(can be done in a future PR)